File: brainfuck_interpreters/cpp/src/Interpreter.cpp
#include <vector>
#include <array>
#include <iostream>

#include "Interpreter.hpp"
#include "Parser.hpp"

Interpreter::Interpreter() :
    _instructions{}, _data{0}, _dp{_data.begin()},
    _ip{_instructions.cbegin()}, _begin{_instructions.cend()},
    _end{_instructions.cend()} {
}

Interpreter::Interpreter(std::vector<Token> instructions) :
    _instructions{instructions}, _data{0}, _dp{_data.begin()},
    _ip{_instructions.cbegin()}, _begin{_instructions.cend()},
    _end{_instructions.cend()} {
}


Interpreter::~Interpreter() {
}

void Interpreter::clean() {
    _instructions = {};
    _data = {0};
    _dp = {};
    _ip = {};
    _begin = {};
    _end = {};
}
void Interpreter::setInstructions(std::vector<Token> instructions) {
    _instructions = {instructions};
}

void Interpreter::run() {
    auto d_end = _data.cend();
    auto d_begin = _data.cbegin();

    while (_ip != _end) {
        switch (*_ip) {
        case MoveFwd:
            if (_dp != d_end) {
                _dp += 1;
            }
            break;
        case MoveBack:
            if (_dp != d_begin) {
                _dp -= 1;
            }
            break;
        case Incr:
            *_dp += 1;
            break;
        case Decr:
            *_dp -= 1;
            break;
        case Print:
            std::cout << *_dp;
            break;
        case Read:
            std::cin >> *_dp;
            break;
        case WhileStart:
            if (*_dp == 0) {
                _jumpForward();
            }
            break;
        case WhileEnd:
            if (*_dp != 0) {
                _jumpBackward();
            }
            break;
        default:
            break;
        }
        _ip += 1;
    }
}

void Interpreter::_jumpForward() {
    int while_level = 0;

    while (true && _ip != _end) {
        if (*_ip == WhileStart) {
            while_level += 1;
        } else if (*_ip == WhileEnd) {
            while_level -= 1;
            if (while_level == 0) {
                break;
            }
        }
        _ip += 1;
    }
}

void Interpreter::_jumpBackward() {
    int while_level = 0;

    while (true && _ip != _begin) {
        if (*_ip == WhileEnd) {
            while_level += 1;
        } else if (*_ip == WhileStart) {
            while_level -= 1;
            if (while_level == 0) {
                break;
            }
        }
        _ip -= 1;
    }
}

File: brainfuck_interpreters/rust/src/interpreter.rs
use std::io;

use parser;

pub struct Interpreter {
    instructions: Vec<parser::Token>,
    data: [i8; 30000],
    ip: usize,
    dp: usize
}

impl Interpreter {
    pub fn new(instructions: Vec<parser::Token>) -> Interpreter {
        return Interpreter{
            instructions: instructions,
            data: [0; 30000],
            ip: 0,
            dp: 0
        };
    }

    pub fn run(&mut self) {
        loop {
            match self.instructions[self.ip] {
                parser::Token::MoveFwd => self.dp += 1,
                parser::Token::MoveBack => self.dp -= 1,
                parser::Token::Incr => self.data[self.dp] += 1,
                parser::Token::Decr => self.data[self.dp] -= 1,
                parser::Token::Print => print!("{}", self.data[self.dp] as u8 as char),
                parser::Token::Read => {
                    let mut input = String::new();
                    io::stdin().read_line(&mut input).unwrap();
                    match input.trim_right().parse::<i8>() {
                        Ok(i) => self.data[self.dp] = i,
                        Err(e) => panic!("Bad value: {}", e)
                    }
                },
                parser::Token::WhileStart => {
                    if self.data[self.dp] == 0 {
                        self.jump_forward()
                    }
                },
                parser::Token::WhileEnd => {
                    if self.data[self.dp] != 0 {
                        self.jump_backward()
                    }
                },
                parser::Token::EOF => return,
                _ => {},
            }
            self.ip += 1;
        }
    }

    fn jump_forward(&mut self) {
        let mut while_level = 0;

        loop {
            if self.instructions[self.ip] == parser::Token::WhileStart {
                while_level += 1;
            } else if self.instructions[self.ip] == parser::Token::WhileEnd {
                while_level -= 1;
                if while_level == 0 {
                    return;
                }
            }
            self.ip += 1;
        }
    }

    fn jump_backward(&mut self) {
        let mut while_level = 0;

        loop {
            if self.instructions[self.ip] == parser::Token::WhileEnd {
                while_level += 1;
            } else if self.instructions[self.ip] == parser::Token::WhileStart {
                while_level -= 1;
                if while_level == 0 {
                    return;
                }
            }
            self.ip -= 1;
        }
    }
}
